app/fetch-did-web-doc.py

```diff
@@ -119,14 +123,120 @@ def request(url, data=None, headers=None):
     }
 
 
+def get_queue_dir(url):
+    queue_id = hashlib.sha256(url.encode("utf-8")).hexdigest()
+    queue_dir = os.path.join(tempfile.gettempdir(), f"scitt-fetch-queue-{queue_id}")
+    return queue_dir
+
+
+def queue_request(url: str, nonce: str, callback_url: str, unattested: bool):
+    logging.info(
+        f"Queuing request for {url} (nonce: {nonce}, callback: {callback_url}, unattested: {unattested})"
+    )
+
+    queue_dir = get_queue_dir(url)
+    request_metadata = {
+        "url": url,
+        "nonce": nonce,
+        "callback_url": callback_url,
+        "unattested": unattested,
+    }
+    request_id = uuid.uuid4().hex
+    request_path = os.path.join(queue_dir, f"{request_id}.json")
+
+    # Write the request metadata to a temporary file.
+    with tempfile.NamedTemporaryFile("w", delete=False) as f:
+        json.dump(request_metadata, f)
+
+    try:
+        while True:
+            # Create the queue folder if it doesn't exist yet.
+            try:
+                os.mkdir(queue_dir)
+            except OSError as e:
+                if e.errno == errno.EEXIST:
+                    is_first = False
+                else:
+                    raise e
+            else:
+                logging.info(f"Created queue folder {queue_dir}")
+                is_first = True
+
+            # Move the request file to the queue folder.
+            try:
+                os.rename(f.name, request_path)
+            except OSError as e:
+                # If the queue folder was deleted in the meantime, try again.
+                if e.errno == errno.ENOENT:
+                    continue
+                else:
+                    raise e
+            else:
+                logging.info(f"Queued request {request_path}")
+            break
+    except Exception as e:
+        logging.error(f"Error while queuing request: {e}")
+        os.remove(f.name)
+        raise e
+
+    return is_first
+
+
+def process_requests(url):
+    queue_dir = get_queue_dir(url)
+
+    # TODO: what if this fails? more entries will be queued but never processed
```

Comment:
One problem with renaming the queue directory is that there will be many more individual queues and hence more resolutions. The test you wrote that counts the server requests would then be either 1 or 2, probably mostly 2. I'm not sure how valid the concern about the continuous stream is, since one of the resolutions would eventually finish and pretty quickly the stream would stop as the resolution would be cached.
But yes, the main problem is leaving ownerless queues behind. Another approach could be to read the queue folder creation time and remove old queues (by moving + removing).

Reply:
This is done now. I previously understood that you wanted to move the queue folder immediately, but we clarified that it should be moved after the first request in a queue got processed which makes more sense.